### PR TITLE
Channels last data transformation and operator wrapping

### PR DESCRIPTION
**Pull request by HenniOVP**
Hi all,
these are the channels last transformations and wrapped operations, which I have previously been working on in `finn-base`.

Similarly to PR #4  this PR also depends on the `qonnx_quant_op` branch of `finn-base`. Additionally, I pushed a change to this branch to enable the execution for qonnx nodes: https://github.com/Xilinx/finn-base/commit/f913744471e2b905b73dc0d7d3777e3538baca87
As such this first version should be equivalent to the branch which I had previously worked on: https://github.com/Xilinx/finn-base/tree/feature/nhwc_ops_and_trafo 

I'll continue working on the PR to implement unit tests, but there are already a few things, which we might want to discuss:

- Naming: Currently I stuck to naming most things `nhwc`, e.g. the transformation combining all other transformations can be imported like this: 
    ```python 
    from qonnx.transformation.nhwc import ConvertToNHWCAndClean
    ```
    and the node execution wrapper is called `NhwcWrappedOp` and lives in `qonnx.custom_op.nhwc.wrapped_ops`. However, if we decide to support 1D operations as well (which we likely want to do), then the term `nhwc` is somewhat misleading, since it explicitly names a 4D-tensor. Maybe naming everything `ChannelsLast` or `ChanLast` would be more accurate?
- What are other pipe-cleaner networks, which I should make sure are working before merging?
   - One network, which I'm already planning to look at is the 1D-Conv network, created for the ITU challenge: https://github.com/Xilinx/brevitas-radioml-challenge-21/blob/main/sandbox/notebooks/models/pretrained_VGG10_w8a8_20_export.onnx

### CODE REVIEW

**Comment by HenniOVP** [comment]
All done. The PR is now ready for review and merging.

File: setup.cfg [hunk 1]
```diff
@@ -64,6 +67,8 @@ testing =
     pytest-cov
 
 [options.entry_points]
+console_scripts =
```
Comment (maltanar) [inline setup.cfg:70]: extra: adding a console script for calling the channels-last transformation would also be useful

File: src/qonnx/custom_op/ChannelsLast/__init__.py [hunk 2]
```diff
@@ -0,0 +1,7 @@
+from qonnx.custom_op.ChannelsLast.wrapped_ops import BatchNormalization, Conv, MaxPool
```
Comment (maltanar) [inline src/qonnx/custom_op/ChannelsLast/__init__.py:1]: to have some consistency for module names I suggest we follow the recommendations from PEP-8 https://www.python.org/dev/peps/pep-0008/#package-and-module-names 

> Modules should have short, all-lowercase names. Underscores can be used in the module name if it improves readability. Python packages should also have short, all-lowercase names, although the use of underscores is discouraged.

such that e.g. `from qonnx.custom_op.ChannelsLast.wrapped_ops import BatchNormalization, Conv, MaxPool` becomes `from qonnx.custom_op.channels_last.wrapped_ops import BatchNormalization, Conv, MaxPool`

should be checked for other module names too

File: src/qonnx/custom_op/ChannelsLast/wrapped_ops.py [hunk 3]
```diff
@@ -0,0 +1,488 @@
...
+        sess = rt.InferenceSession(intermediate_model.SerializeToString())
+        output_list = sess.run(None, input_dict)
+        output_onnx = output_list[0]
+
+        # Transpose the output back to channel last and save it in the external context.
+        output_onnx = output_onnx.transpose(self._to_chan_last_args[ndim])
+        context[node.output[0]] = output_onnx
+
+
+class Conv(ChannelsLastWrappedOp):
```
Comment (maltanar) [inline src/qonnx/custom_op/ChannelsLast/wrapped_ops.py:101]: suggest to move each instances of `ChannelsLastWrapperOp` into its own module (so e.g. this one goes under `channelslast/conv.py`

File: src/qonnx/custom_op/ChannelsLast/wrapped_ops.py [hunk 4]
```diff
@@ -0,0 +1,488 @@
...
+    # Transpose parameters to convert to channels last for 3D and 4D tensors
+    _to_chan_last_args = {
+        3: (0, 2, 1),
+        4: (0, 2, 3, 1),
+    }
+    # Similarly for converting back to channels first.
+    _to_chan_first_args = {
+        3: (0, 2, 1),
+        4: (0, 3, 1, 2),
+    }
```
Comment (maltanar) [inline src/qonnx/custom_op/ChannelsLast/wrapped_ops.py:26]: consider making a utility function to return the axis permutation order (both ways, chanfirst -> chanlast and chanlast -> chanfirst) given the number of dimensions

chanfirst -> chanlast : create list(range(ndim)) then move the dimension with ind "1" to the final position
chanlast -> chanfirst : create list(range(ndim)) then move the dimension with ind "ndim-1" to the next-first position

File: src/qonnx/custom_op/ChannelsLast/wrapped_ops.py [hunk 5]
```diff
@@ -0,0 +1,488 @@
...
+
+        if not verification_successful:
+            raise RuntimeError(
+                f"Verification of node {node.name} failed, please check the " f"attached info messages: {info_messages}"
+            )
+
+        return info_messages
+
+
+class MaxPool(ChannelsLastWrappedOp):
```
Comment (maltanar) [inline src/qonnx/custom_op/ChannelsLast/wrapped_ops.py:249]: suggest to move to own module, see comments about Conv

File: tests/transformations/test_ChannelsLast.py [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+    elif "RadioML_VGG10":
+        input_shape = (1, 2, 1024)
+    else:
+        raise ValueError(f"Model called {test_model} is not supported")
+    input_tensor = np.ones(np.prod(np.asarray(input_shape)), dtype=np.float32)
+    input_tensor = input_tensor.reshape(input_shape)
+
+    model = ModelWrapper(onnx_file)
+    input_dict = {model.graph.input[0].name: input_tensor}
+    golden_output_dict = oxe.execute_onnx(model, input_dict, True)
```
Comment (maltanar) [inline tests/transformations/test_ChannelsLast.py:58]: do you need the `True` here? (returns full execution context)

File: tests/transformations/test_ChannelsLast.py [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+    result = dict()
+    for n in model.graph.node:
+        if is_finn_op(n.domain):
+            n_instance = getCustomOp(n)
+            verify_result = n_instance.verify_node()
+            result[n.name] = verify_result
+    return result
+
+
+def analysis_first_node_is_transpose(model):
```
Comment (maltanar) [inline tests/transformations/test_ChannelsLast.py:100]: consider adding an option to also call `MakeInputChannelsLast` as part of the channels-last transformation: https://github.com/Xilinx/finn-base/blob/dev/src/finn/transformation/make_input_chanlast.py#L35

Comment (maltanar) [reply tests/transformations/test_ChannelsLast.py:100]: note that in this case the global input tensor shape itself will change and there should be no more transpose node at the beginning once everything is collapsed

File: src/qonnx/transformation/channelsLast.py [hunk 8]
```diff
@@ -0,0 +1,460 @@
...
+                    continue
+                # Check that this is a "to chan first" trafo
+                perm_1 = get_by_name(n.attribute, "perm")
+                if list(_to_chan_first_args[ndim]) == perm_1.ints:
+
+                    successor_nodes = model.find_direct_successors(n)
+                    assert (
+                        len(successor_nodes) == 1
+                    ), "Transpose nodes should only have one output, I don't think more than one would even be possible."
+                    successor_node = successor_nodes[0]
```
Comment (thesps) [inline src/qonnx/transformation/channelsLast.py:190]: If the `Transpose` node was one inserted by `InsertChannelsLastDomainsAndTrafos`, and the node it was inserted after had multiple successors (e.g. a model with branches), could the `Transpose` node have multiple successors?

File: src/qonnx/transformation/channelsLast.py [hunk 9]
```diff
@@ -0,0 +1,460 @@
...
+    Converts data layout dependent nodes to ChannelsLast nodes and inserts transformations.
+    Then it tries to eliminate as many transformations as possible and moves the
+    still existing ones as far upstream as possible.
+    """
+
+    def apply(self, model):
+        assert model.check_all_tensor_shapes_specified(), "All tensor shapes should be specified."
+        model = model.transform(InsertChannelsLastDomainsAndTrafos())
+        max_tries = 10
+        for i in range(max_tries):
```
Comment (maltanar) [inline src/qonnx/transformation/channelsLast.py:45]: make the 10 max_tries here an optional variable to the transformation's initializer function, stored as a member variable

ideally this should not be needed at all (see comment at the end of this function)

**Review by maltanar — CHANGES_REQUESTED** [review]

File: src/qonnx/transformation/channelsLast.py (hunk 8 above)
Comment (HenniOVP) [reply src/qonnx/transformation/channelsLast.py:190]: That was my thinking, but after taking a look at a ResNet model ( https://github.com/onnx/models/blob/master/vision/classification/resnet/model/resnet50-v1-12.onnx ) it seems that this is shouldn't be of concern. Because a branch connection is implemented by connecting the same tensor to multiple nodes. This would then mean that the Transpose node still only has one output. The documentation also defines no optional inputs or outputs: https://github.com/onnx/onnx/blob/master/docs/Operators.md#Transpose
So I think it should be save to remove this statement.
An issue we could run into in the future is that people might use the optional outputs of the MaxPool and BatchNorm nodes, see: https://github.com/onnx/onnx/blob/master/docs/Operators.md#MaxPool and https://github.com/onnx/onnx/blob/master/docs/Operators.md#BatchNormalization
But at least for BatchNorm they are only used for training, which we don't support anyways. And I've also never seen a model make use of the optional outputs.

File: src/qonnx/transformation/channelsLast.py [hunk 10]
```diff
@@ -0,0 +1,460 @@
...
+            # Check if the model changed
+            new_model_string = model.model.SerializeToString()
+            if not (initial_model_string == new_model_string):
+                # Do some cleanup
+                model = model.transform(RemoveUnusedTensors())
+                model = model.transform(InferShapes())
+            else:
+                break
+
+        return model, False
```
[the remark this answers is not in the document]
Comment (HenniOVP) [reply src/qonnx/transformation/channelsLast.py:76]: Hm, not really, I also haven't seen it happen during development.

File: src/qonnx/transformation/channelsLast.py (hunk 8 above)
Comment (HenniOVP) [reply src/qonnx/transformation/channelsLast.py:190]: Hi,
I'll give the ResNet ONNX file from the ONNX model zoo a try and check what happens.

File: src/qonnx/custom_op/ChannelsLast/__init__.py (hunk 2 above)
Comment (HenniOVP) [reply src/qonnx/custom_op/ChannelsLast/__init__.py]: This should be addressed now.

File: src/qonnx/custom_op/ChannelsLast/wrapped_ops.py (hunk 3 above)
Comment (HenniOVP) [reply src/qonnx/custom_op/ChannelsLast/wrapped_ops.py]: Okay, done. This also makes the structure in files a lot cleaner.

File: src/qonnx/custom_op/ChannelsLast/wrapped_ops.py (hunk 5 above)
Comment (HenniOVP) [reply src/qonnx/custom_op/ChannelsLast/wrapped_ops.py]: Done

File: src/qonnx/custom_op/ChannelsLast/wrapped_ops.py [hunk 11]
```diff
@@ -0,0 +1,488 @@
...
+
+        if not verification_successful:
+            raise RuntimeError(
+                f"Verification of node {node.name} failed, please check the " f"attached info messages: {info_messages}"
+            )
+
+        return info_messages
+
+
+class BatchNormalization(ChannelsLastWrappedOp):
```
[the remark this answers is not in the document]
Comment (HenniOVP) [reply src/qonnx/custom_op/ChannelsLast/wrapped_ops.py]: Done

File: tests/transformations/test_ChannelsLast.py [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+
+
+def get_golden_in_and_output(onnx_file, test_model):
+    if test_model == "FINN-CNV_W2A2":
+        input_shape = (1, 3, 32, 32)
+    elif "RadioML_VGG10":
+        input_shape = (1, 2, 1024)
+    else:
+        raise ValueError(f"Model called {test_model} is not supported")
+    input_tensor = np.ones(np.prod(np.asarray(input_shape)), dtype=np.float32)
```
[the remark this answers is not in the document]
Comment (HenniOVP) [reply tests/transformations/test_ChannelsLast.py]: Okay, I adjusted it accordingly. Please take another look if the value range is okay.

File: tests/transformations/test_ChannelsLast.py (hunk 7 above)
Comment (HenniOVP) [reply tests/transformations/test_ChannelsLast.py:111]: Yeah, it now adds two test cases, but I think that's fine.

File: src/qonnx/transformation/channelsLast.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (HenniOVP) [reply src/qonnx/transformation/channelsLast.py]: Okay, it's now implemented like this and it works quite nicely.

File: src/qonnx/custom_op/ChannelsLast/wrapped_ops.py (hunk 4 above)
Comment (HenniOVP) [reply src/qonnx/custom_op/ChannelsLast/wrapped_ops.py]: Okay, I moved this into functions in the `base_wrapped_op.py` file.

File: tests/transformations/test_ChannelsLast.py (hunk 6 above)
Comment (HenniOVP) [reply tests/transformations/test_ChannelsLast.py]: Yeah, it's not really needed, so I removed it.

File: src/qonnx/transformation/QuantConstantFolding.py [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+                predecessors = model.find_direct_predecessors(n)
+                # Check if we reached the top of the graph
+                if predecessors is None:
+                    continue
+                assert (
+                    len(predecessors) == 1
+                ), "Transpose nodes should only have one input, I don't think more than one would even be possible."
+                predecessor = predecessors[0]
+                if predecessor.op_type == "Quant":
+                    inp = predecessor.input[0]
```
[the remark this answers is not in the document]
Comment (HenniOVP) [reply src/qonnx/transformation/QuantConstantFolding.py]: True, I adjusted the transformation to also transpose initializes, which fit the shape of the transposes parameters, so it doesn't try to transpose scalars.
I've also removed duplicated code, which did basically the same in the `MoveChanLastUpstream` transformation and instead I now call the `FoldTransposeIntoQuantInit` afterwards to achieve the same effect.

File: setup.cfg (hunk 1 above)
Comment (HenniOVP) [reply setup.cfg:70]: Yeah, that is nice to have. I added support for it in the most recent push.

File: src/qonnx/transformation/channelsLast.py (hunk 8 above)
Comment (HenniOVP) [reply src/qonnx/transformation/channelsLast.py]: Okay, adding support for this appears to be a bit more involved. So I've added a check a the beginning of the transformation to check that a model is linear before it gets processed.

**Comment by maltanar** [comment]
Just FYI, I pushed some extra commits to:
* add TFC as well to the list of channels-last tests (basically nothing should happen)
* refactored the channels-last tests to keep details in a map, easier to add new tests this way
* covered some corner cases in channels-last transformations (@HenniOVP remember that `ModelWrapper.find_*` functions can return `None` and this must always be checked)
